src/daggy.js

```diff
@@ -1,127 +1,151 @@
-  Option.None instanceof Option; // true
-
-  function incOrZero(o) {
-      return o.cata({
-          Some: function(x) {
-              return x + 1;
-          },
-          None: function() {
-              return 0;
-          }
-      });
+const taggedSum$cata = function (fs) {
+  const tag = this[TAG]
+  if (!fs[tag]) {
+    throw new TypeError("Constructors given to cata didn't include: " + tag)
   }
-  incOrZero(Option.Some(1)); // 2
-  incOrZero(Option.None); // 0
-  ```
-**/
-function taggedSum(constructors) {
-    var key,
-        ctor;
-
-    function definitions() {
-        throw new TypeError('Tagged sum was called instead of one of its properties.');
-    }
+  return fs[tag].apply(fs, this[VALUES])
+}
 
-    function makeCata(key) {
-        // Note: we need the prototype from this function.
-        return function(dispatches) {
-            var i;
+const taggedSum$makeValue = (typeRep, tag, values) => {
+  const fields = typeRep['@@union'][tag]
+  if (values.length !== fields.length) {
+    throw new TypeError(`Expected ${fields.length} arguments, got ${values.length}`)
+  }
+  const obj = Object.create(typeRep.prototype)
+  defProp(obj, VALUES, values)
+  defProp(obj, TAG, tag)
+  for (let idx = 0; idx < fields.length; idx++) {
+    obj[fields[idx]] = values[idx]
+  }
+  return obj
+}
 
-            const fields = constructors[key];
-            const args = [];
+const taggedSum$constructorToString = function () {
+  return `${this['@@typeRep']['@@type']}.${this[TAG]}`
+}
 
-            if(!dispatches[key])
-                throw new TypeError("Constructors given to cata didn't include: " + key);
+const typeRepToString = function () {
+  return this['@@type']
+}
 
-            for(i = 0; i < fields.length; i++)
-                args.push(this[fields[i]]);
+const taggedSum$objToString = function () {
+  return `${this.constructor['@@type']}.${this[TAG]}(${arrToString(this[VALUES])})`
+}
 
-            return dispatches[key].apply(this, args);
-        };
-    }
+const taggedSum$isVariant = function (val) {
+  return Boolean(val) &&
+    this[TAG] === val[TAG] &&
+    this['@@typeRep']['@@type'] === type(val)
+}
 
-    function makeProto(key) {
-        const proto = Object.create(definitions.prototype);
-        proto.cata = makeCata(key);
-        return proto;
-    }
+const taggedSum$isUnitValue = function (val) {
+  return this === val || Boolean(val) &&
+    this[TAG] === val[TAG] &&
+    type(this) === type(val)
+}
 
-    for(key in constructors) {
-        if(!constructors[key].length) {
-            definitions[key] = makeProto(key);
-            definitions[key].toString = constant('()');
-            continue;
-        }
-        ctor = tagged.apply(null, constructors[key]);
-        definitions[key] = ctor;
-        definitions[key].prototype = makeProto(key);
-        definitions[key].prototype.constructor = ctor;
-    }
+const isType = function (val) {
+  return Boolean(val) &&
```

Comment:
If this function would be given a falsy value like `undefined`, it would return it right back. So the type of this function is `isType :: Any -> Boolean|Null|undefined|String|Number` unless explicitly cast to Boolean.